plugin/trino-hive/src/main/java/io/trino/plugin/hive/fs/HiveFileIterator.java

```diff
@@ -75,63 +79,102 @@ public HiveFileIterator(
+            }
+            else if (isHiddenFileOrDirectory(status.getPath())) {
+                continue;
             }
 
-            if (paths.isEmpty()) {
-                return endOfData();
+            if (status.isDirectory()) {
+                switch (nestedDirectoryPolicy) {
+                    case IGNORED:
+                        continue;
+                    case RECURSE:
+                        // Recursive listings call listFiles which should not return directories, this is a contract violation
+                        // and can be handled the same way as the FAIL case
+                    case FAIL:
+                        throw new NestedDirectoryNotAllowedException(status.getPath());
+                }
             }
-            remoteIterator = getLocatedFileStatusRemoteIterator(paths.removeFirst());
+            return status;
         }
+
+        return endOfData();
     }
 
     private Iterator<LocatedFileStatus> getLocatedFileStatusRemoteIterator(Path path)
     {
         try (TimeStat.BlockTimer ignored = namenodeStats.getListLocatedStatus().time()) {
-            if (ignoreAbsentPartitions && !exists(path)) {
-                return emptyIterator();
+            return new FileStatusIterator(table, path, fileSystem, directoryLister, namenodeStats, nestedDirectoryPolicy == NestedDirectoryPolicy.RECURSE);
+        }
+        catch (TrinoException e) {
+            if (ignoreAbsentPartitions) {
+                try {
+                    if (!fileSystem.exists(path)) {
+                        return emptyIterator();
+                    }
+                }
+                catch (Exception ee) {
+                    TrinoException trinoException = new TrinoException(HIVE_FILESYSTEM_ERROR, "Failed to check if path exists: " + path, ee);
+                    trinoException.addSuppressed(e);
+                    throw trinoException;
+                }
             }
-            return new FileStatusIterator(table, path, fileSystem, directoryLister, namenodeStats);
+            throw e;
         }
     }
 
-    private boolean exists(Path path)
+    private LocatedFileStatus getLocatedFileStatus(Iterator<LocatedFileStatus> iterator)
     {
-        try {
-            return fileSystem.exists(path);
-        }
-        catch (IOException e) {
-            throw new TrinoException(HIVE_FILESYSTEM_ERROR, "Failed to check if path exists: " + path, e);
+        try (TimeStat.BlockTimer ignored = namenodeStats.getRemoteIteratorNext().time()) {
+            return iterator.next();
         }
     }
 
-    private LocatedFileStatus getLocatedFileStatus(Iterator<LocatedFileStatus> iterator)
+    @VisibleForTesting
+    static boolean isHiddenFileOrDirectory(Path path)
     {
-        try (TimeStat.BlockTimer ignored = namenodeStats.getRemoteIteratorNext().time()) {
-            return iterator.next();
+        // Only looks for the last part of the path
+        String pathString = path.toUri().getPath();
```

Comment:
why not simply `path.getName()`?

Reply:
If you look at the logic in `Path#getName()`, it's doing this exact thing, except it returns the sub-string (a new allocation). This seemed like an easier pattern to follow to make these methods share code with a smaller footprint.